File: services/mobile-client/lib/navigation/navigation_math.dart
import 'dart:math' as math;

import 'package:latlong2/latlong.dart';

class RouteProximity {
  const RouteProximity({required this.distanceM, required this.segmentIndex});

  final double distanceM;
  final int segmentIndex;
}

class RerouteDecision {
  const RerouteDecision({
    required this.trigger,
    required this.offRouteSince,
    required this.effectiveThresholdM,
  });

  final bool trigger;
  final DateTime? offRouteSince;
  final double effectiveThresholdM;
}

class NavigationMath {
  static const double _earthRadiusM = 6371000;

  static double normalizeHeading(double value) => (value % 360 + 360) % 360;

  static RouteProximity closestRouteSegment(
    LatLng point,
    List<LatLng> route,
  ) {
    if (route.isEmpty) {
      return const RouteProximity(
        distanceM: double.infinity,
        segmentIndex: 0,
      );
    }
    if (route.length == 1) {
      return RouteProximity(
        distanceM: pointToSegmentDistanceM(point, route.first, route.first),
        segmentIndex: 0,
      );
    }
    var bestDistance = double.infinity;
    var bestIndex = 0;
    for (var index = 0; index < route.length - 1; index++) {
      final distance = pointToSegmentDistanceM(
        point,
        route[index],
        route[index + 1],
      );
      if (distance < bestDistance) {
        bestDistance = distance;
        bestIndex = index;
      }
    }
    return RouteProximity(distanceM: bestDistance, segmentIndex: bestIndex);
  }

  static double pointToSegmentDistanceM(
    LatLng point,
    LatLng start,
    LatLng end,
  ) {
    final referenceLat =
        (point.latitude + start.latitude + end.latitude) / 3;
    final latitudeScale = math.pi * _earthRadiusM / 180;
    final longitudeScale = latitudeScale * math.cos(referenceLat * math.pi / 180);

    ({double x, double y}) project(LatLng value) => (
          x: value.longitude * longitudeScale,
          y: value.latitude * latitudeScale,
        );

    final p = project(point);
    final a = project(start);
    final b = project(end);
    final dx = b.x - a.x;
    final dy = b.y - a.y;
    final lengthSquared = dx * dx + dy * dy;
    if (lengthSquared == 0) return math.sqrt(math.pow(p.x - a.x, 2) + math.pow(p.y - a.y, 2));
    final rawT = ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared;
    final t = rawT.clamp(0.0, 1.0);
    final nearestX = a.x + t * dx;
    final nearestY = a.y + t * dy;
    return math.sqrt(
      math.pow(p.x - nearestX, 2) + math.pow(p.y - nearestY, 2),
    );
  }

  static RerouteDecision shouldReroute({
    required double distanceM,
    required double accuracyM,
    required DateTime now,
    required DateTime? offRouteSince,
    required DateTime? lastRerouteAt,
    required bool rerouting,
    double thresholdM = 60,
    Duration confirmation = const Duration(seconds: 4),
    Duration cooldown = const Duration(seconds: 20),
  }) {
    final effectiveThreshold = math.max(thresholdM, math.max(0, accuracyM) * 2);
    if (!distanceM.isFinite || distanceM <= effectiveThreshold) {
      return RerouteDecision(
        trigger: false,
        offRouteSince: null,
        effectiveThresholdM: effectiveThreshold,
      );
    }
    final startedAt = offRouteSince ?? now;
    final confirmed = now.difference(startedAt) >= confirmation;
    final cooledDown = lastRerouteAt == null || now.difference(lastRerouteAt) >= cooldown;
    return RerouteDecision(
      trigger: confirmed && cooledDown && !rerouting,
      offRouteSince: startedAt,
      effectiveThresholdM: effectiveThreshold,
    );
  }
}

File: backend/presentation/static/navigation_math.js
(function (root, factory) {
  const api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  root.MapNetNavigationMath = api;
})(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  const EARTH_RADIUS_M = 6371000;

  function toRadians(value) {
    return value * Math.PI / 180;
  }

  function normalizeHeading(value) {
    if (!Number.isFinite(value)) return null;
    return ((value % 360) + 360) % 360;
  }

  function projectMeters(point, referenceLat) {
    const latitudeScale = Math.PI * EARTH_RADIUS_M / 180;
    const longitudeScale = latitudeScale * Math.cos(toRadians(referenceLat));
    return {
      x: point.lng * longitudeScale,
      y: point.lat * latitudeScale,
    };
  }

  function pointToSegmentDistanceMeters(point, start, end) {
    const referenceLat = (point.lat + start.lat + end.lat) / 3;
    const p = projectMeters(point, referenceLat);
    const a = projectMeters(start, referenceLat);
    const b = projectMeters(end, referenceLat);
    const dx = b.x - a.x;
    const dy = b.y - a.y;
    const lengthSquared = dx * dx + dy * dy;
    if (lengthSquared === 0) return Math.hypot(p.x - a.x, p.y - a.y);
    const t = Math.max(0, Math.min(1, ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared));
    return Math.hypot(p.x - (a.x + t * dx), p.y - (a.y + t * dy));
  }

  function closestRouteSegment(point, coordinates) {
    if (!Array.isArray(coordinates) || coordinates.length === 0) {
      return { distance: Infinity, segmentIndex: 0 };
    }
    if (coordinates.length === 1) {
      const only = { lng: coordinates[0][0], lat: coordinates[0][1] };
      return {
        distance: pointToSegmentDistanceMeters(point, only, only),
        segmentIndex: 0,
      };
    }
    let bestDistance = Infinity;
    let segmentIndex = 0;
    for (let index = 0; index < coordinates.length - 1; index += 1) {
      const distance = pointToSegmentDistanceMeters(
        point,
        { lng: coordinates[index][0], lat: coordinates[index][1] },
        { lng: coordinates[index + 1][0], lat: coordinates[index + 1][1] },
      );
      if (distance < bestDistance) {
        bestDistance = distance;
        segmentIndex = index;
      }
    }
    return { distance: bestDistance, segmentIndex };
  }

  function shouldReroute(options) {
    const {
      distance,
      accuracy = 0,
      now,
      offRouteSince,
      lastRerouteAt = 0,
      threshold = 80,
      confirmationMs = 4000,
      cooldownMs = 20000,
      rerouting = false,
    } = options;
    const effectiveThreshold = Math.max(threshold, Math.max(0, accuracy) * 2);
    if (!Number.isFinite(distance) || distance <= effectiveThreshold) {
      return { trigger: false, offRouteSince: null, effectiveThreshold };
    }
    const startedAt = offRouteSince == null ? now : offRouteSince;
    const confirmed = now - startedAt >= confirmationMs;
    const cooledDown = now - lastRerouteAt >= cooldownMs;
    return {
      trigger: confirmed && cooledDown && !rerouting,
      offRouteSince: startedAt,
      effectiveThreshold,
    };
  }

  return {
    normalizeHeading,
    pointToSegmentDistanceMeters,
    closestRouteSegment,
    shouldReroute,
  };
});
